File: kotlin/03-operators/Operators.kt
fun main() {
    // Arithmetic operators
    val a = 17
    val b = 5
    println("a + b = ${a + b}")
    println("a - b = ${a - b}")
    println("a * b = ${a * b}")
    println("a / b = ${a / b}")    // Integer division
    println("a % b = ${a % b}")    // Remainder

    // Floating-point division
    val x = 17.0
    val y = 5.0
    println("x / y = ${x / y}")

    // Compound assignment
    var counter = 10
    counter += 5
    counter -= 2
    counter *= 2
    println("counter = $counter")

    // Increment and decrement
    var n = 1
    n++
    ++n
    println("n = $n")

    // Comparison operators
    println("a == b is ${a == b}")
    println("a != b is ${a != b}")
    println("a > b is ${a > b}")
    println("a <= b is ${a <= b}")

    // Logical operators (short-circuit)
    val isAdult = true
    val hasTicket = false
    println("isAdult && hasTicket = ${isAdult && hasTicket}")
    println("isAdult || hasTicket = ${isAdult || hasTicket}")
    println("!isAdult = ${!isAdult}")

    // String concatenation with +
    val greeting = "Hello, " + "Kotlin!"
    println(greeting)

    // Bitwise operations use infix functions, not symbols
    val flags = 0b1100 and 0b1010
    val mask = 0b1100 or 0b0011
    val shifted = 1 shl 3
    println("and = $flags, or = $mask, shl = $shifted")
}

File: rust/03-operators/operators.rs
fn main() {
    // --- Arithmetic operators ---
    let a = 17;
    let b = 5;
    println!("Arithmetic:");
    println!("  {} + {} = {}", a, b, a + b);
    println!("  {} - {} = {}", a, b, a - b);
    println!("  {} * {} = {}", a, b, a * b);
    println!("  {} / {} = {}", a, b, a / b); // integer division truncates
    println!("  {} % {} = {}", a, b, a % b); // remainder

    // Floating-point division keeps the fraction
    let x = 17.0;
    let y = 5.0;
    println!("  {} / {} = {}", x, y, x / y);

    // --- Comparison operators (always produce a bool) ---
    println!("Comparison:");
    println!("  {} == {} -> {}", a, b, a == b);
    println!("  {} != {} -> {}", a, b, a != b);
    println!("  {} <  {} -> {}", a, b, a < b);
    println!("  {} >= {} -> {}", a, b, a >= b);

    // --- Logical operators (short-circuiting) ---
    let sunny = true;
    let warm = false;
    println!("Logical:");
    println!("  sunny && warm -> {}", sunny && warm);
    println!("  sunny || warm -> {}", sunny || warm);
    println!("  !sunny        -> {}", !sunny);

    // --- Compound assignment (requires `mut`) ---
    let mut total = 10;
    total += 5;
    total -= 3;
    total *= 2;
    println!("Compound assignment: total = {}", total);

    // --- Bitwise operators ---
    let flags = 0b1100;
    let mask = 0b1010;
    println!("Bitwise:");
    println!("  {:04b} & {:04b} = {:04b}", flags, mask, flags & mask);
    println!("  {:04b} | {:04b} = {:04b}", flags, mask, flags | mask);
    println!("  {:04b} ^ {:04b} = {:04b}", flags, mask, flags ^ mask);
    println!("  1 << 4 = {}", 1 << 4);

    // --- Operator precedence: * and / bind tighter than + and - ---
    println!("Precedence:");
    println!("  2 + 3 * 4   = {}", 2 + 3 * 4);
    println!("  (2 + 3) * 4 = {}", (2 + 3) * 4);

    // --- Range operator (..= is inclusive) ---
    let sum: i32 = (1..=5).sum();
    println!("Range: sum of 1..=5 = {}", sum);
}
